### PR TITLE
jellyfish-block implementation

### LINKED ISSUE #490: `jellyfish-block` implementation
#### What would you like to be added:

Similar to `jellyfish-transaction`, this package implements `Block` and `CBlock` with BufferComposer.

#### Why is this needed:

- For better indexing interoperability with defid.
- Size and performance conscious index implementation

/area core
/triage accepted

### PR DESCRIPTION
**Pull request by Jouzo**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:

Creates new `jellyfish-block` package. 
Buffer composer for Block and CBlock.

Needed for :

   -  better indexing interoperability with defid.
   -  performance conscious index implementation

#### Which issue(s) does this PR fixes?:

Fixes #490

### CODE REVIEW

**Comment by jingyi2811** [comment]
So, `blockHeader + extendedTransaction = block`, right? 
I think overall it looks good to me.

**Comment by Jouzo** [comment]
> So, `blockHeader + extendedTransaction = block`, right?
> I think overall it looks good to me.

Yes, BlockHeader + ExtendedTransaction[]

**Comment by jingyi2811** [comment]
There are minor problems in this PR. But I think it is okay because this is a DRAFT PR. You will improve the code as the time goes.

File: packages/jellyfish-block/src/blockHeader.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import { BufferComposer, ComposableBuffer } from '@defichain/jellyfish-buffer'
+import BigNumber from 'bignumber.js'
+
+export interface BlockHeader {
+  version: number // ------------| 4 bytes, block version number
```
Comment (canonbrother) [inline packages/jellyfish-block/src/blockHeader.ts:5]: hmm.. should we follow the ain using `nVersion` ??

File: packages/jellyfish-block/src/block.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  TransactionSegWit,
+  Transaction,
+  CTransaction,
+  CTransactionSegWit
+} from '@defichain/jellyfish-transaction'
+import { SmartBuffer } from 'smart-buffer'
+
+export interface Block {
+  blockHeader: BlockHeader // -------------| 124 bytes + signature (n = VarUInt{1-9 bytes} + n bytes)
+  transactions: Array<TransactionSegWit | Transaction> // ----| n = VarUInt{1-9 bytes}, + n transactions, list of transactions
```
Comment (canonbrother) [inline packages/jellyfish-block/src/block.ts:13]: we need a `Block { blockHeader, transactions<Transaction[]> }` test for document as well

Comment (Jouzo) [reply packages/jellyfish-block/src/block.ts:13]: There is a test block with  `describe('block with transaction as Transaction'` in block.test.ts now.
Is that what you mean ?

Comment (canonbrother) [reply packages/jellyfish-block/src/block.ts:13]: yep.. see that now

**Review by canonbrother — DISMISSED** [review]

File: .github/CODEOWNERS [hunk 3]
```diff
@@ -11,6 +11,8 @@
 
 /packages/jellyfish-api-jsonrpc/                @fuxingloh @thedoublejay
 
+/packages/jellyfish-block/                     @fuxingloh @ivan-zynesis @canonbrother @monstrobishi @surangap @Jouzo @jingyi2811
```
Comment (jingyi2811) [inline .github/CODEOWNERS:14]: Alignment problem

**Review by canonbrother — DISMISSED** [review]

File: packages/jellyfish-block/src/block.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      ComposableBuffer.single<BlockHeader>(() => block.blockHeader, v => block.blockHeader = v, v => new CBlockHeader(v)),
+      ComposableBuffer.varUIntArray<TransactionSegWit | Transaction>(() => block.transactions, v => block.transactions = v, v => {
+        /**
+         * Has to read the marker from buffer in order to determine wether it is a Transaction or TransactionSegWit.
+         */
+        if (v instanceof SmartBuffer) {
+          const readOffset = v.readOffset // Keep current offset in order to reset to this point once marker is read.
+          v.readUInt32LE() // Read 4 bytes from version and discard it.
+          const marker = v.readUInt8()
+          v.readOffset = readOffset
```
Comment (fuxingloh) [inline packages/jellyfish-block/src/block.ts:32]: this is interesting, maybe we should have a `Transaction` automatically map to `SegWit`

**Review by fuxingloh — CHANGES_REQUESTED** [review]
requesting a few more tests that starts a `testcontaienrs`, fetch the `block` from RPC then attempt to parse it and compare with the RPC result

File: packages/jellyfish-block/__tests__/rpc/block.test.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+  })
+
+  afterAll(async () => {
+    await testing.container.stop()
+  })
+
+  /**
+   * Wait for block hash to reach a certain height
+   */
+  async function waitForBlockHash (height: number): Promise<string> {
```
Comment (jingyi2811) [inline packages/jellyfish-block/__tests__/rpc/block.test.ts:22]: You can find 5 duplication of this function code in Jellyfish. Probably you can take the initiative to move this to jellyfish-testing so everyone can reuse the same function

File: packages/jellyfish-block/__tests__/rpc/blockHeader.test.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  })
+
+  afterAll(async () => {
+    await testing.container.stop()
+  })
+
+  /**
+   * Wait for block hash to reach a certain height
+   */
+  async function waitForBlockHash (height: number): Promise<string> {
```
Comment (jingyi2811) [inline packages/jellyfish-block/__tests__/rpc/blockHeader.test.ts:21]: https://github.com/DeFiCh/jellyfish/pull/636/files/8401b0a877187bb70f091196b805cce065abbe22..6545ff4aad121821d2b9e649e9be9f8acbd8f42d#r707411054

File: packages/jellyfish-block/__tests__/rpc/block.test.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    const block = await testing.rpc.blockchain.getBlock(blockHash, 1)
+
+    const composableObject = composable.toObject()
+    expect(composableObject.blockHeader.version).toStrictEqual(block.version)
+    expect(composableObject.blockHeader.hashPrevBlock).toStrictEqual(block.previousblockhash)
+    expect(composableObject.blockHeader.hashMerkleRoot).toStrictEqual(block.merkleroot)
+    expect(composableObject.blockHeader.hashMerkleRoot).toStrictEqual(block.tx[0])
+    expect(composableObject.blockHeader.time).toStrictEqual(block.time)
+    expect(composableObject.blockHeader.bits).toStrictEqual(parseInt(block.bits, 16))
+    expect(composableObject.blockHeader.height).toStrictEqual(new BigNumber(block.height))
```
Comment (jingyi2811) [inline packages/jellyfish-block/__tests__/rpc/block.test.ts:42]:
```suggestion
    expect(composableObject.blockHeader.height).toStrictEqual(new BigNumber(block.height))
    expect(composableObject.blockHeader.stakeModifier).toStrictEqual(block.stakeModifier)
    expect(composableObject.blockHeader.mintedBlocks).toStrictEqual(new BigNumber(block.mintedBlocks))
```

File: packages/jellyfish-block/__tests__/rpc/block.test.ts [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+    const block = await testing.rpc.blockchain.getBlock(blockHash, 2)
+
+    const composableObject = composable.toObject()
+    expect(composableObject.blockHeader.version).toStrictEqual(block.version)
+    expect(composableObject.blockHeader.hashPrevBlock).toStrictEqual(block.previousblockhash)
+    expect(composableObject.blockHeader.hashMerkleRoot).toStrictEqual(block.merkleroot)
+    expect(composableObject.blockHeader.time).toStrictEqual(block.time)
+    expect(composableObject.blockHeader.bits).toStrictEqual(parseInt(block.bits, 16))
+    expect(composableObject.blockHeader.height).toStrictEqual(new BigNumber(block.height))
```
Comment (jingyi2811) [inline packages/jellyfish-block/__tests__/rpc/block.test.ts:61]:
```suggestion
    expect(composableObject.blockHeader.height).toStrictEqual(new BigNumber(block.height))
    expect(composableObject.blockHeader.stakeModifier).toStrictEqual(block.stakeModifier)
    expect(composableObject.blockHeader.mintedBlocks).toStrictEqual(new BigNumber(block.mintedBlocks))
```

File: packages/jellyfish-block/__tests__/rpc/block.test.ts (hunk 7 above)
Comment (Jouzo) [reply packages/jellyfish-block/__tests__/rpc/block.test.ts:42]: The interface from the composer and the one from the RPC don't match exactly. 
The RPC does not send the mintedBlocks nor the stakeModifier.

Comment (surangap) [reply packages/jellyfish-block/__tests__/rpc/block.test.ts:34]: Actually `getblock` RPC returns those fields too. refer https://github.com/DeFiCh/ain/blob/6059de0c533710e7f0214128719201a91230bba1/src/rpc/blockchain.cpp#L148
What it does not return is the `signature` field. not sure why, but better to check with C++ side @Jouzo

File: packages/jellyfish-block/__tests__/rpc/blockHeader.test.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    await testing.container.stop()
+  })
+
+  it('should getBlockHeader with verbosity=true and have same data as CBlockHeader composer to object', async () => {
+    const blockHash = await testing.misc.waitForBlockHash(1)
+    const hexData = await testing.rpc.blockchain.getBlockHeader(blockHash, false)
+    const buffer = SmartBuffer.fromBuffer(Buffer.from(hexData, 'hex'))
+    const composable = new CBlockHeader(buffer)
+
+    const blockHeader = await testing.rpc.blockchain.getBlockHeader(blockHash, true)
```
Comment (surangap) [inline packages/jellyfish-block/__tests__/rpc/blockHeader.test.ts:24]: `getblockheader` RPC does not return `mintedBlocks` and `signature`. should it return those as well? if then log a ticket to c++ side.

File: tsconfig.json [hunk 10]
```diff
@@ -24,6 +25,7 @@
       "@defichain/jellyfish-address/*": ["jellyfish-address/src/*"],
       "@defichain/jellyfish-api-core/*": ["jellyfish-api-core/src/*"],
       "@defichain/jellyfish-api-jsonrpc/*": ["jellyfish-api-jsonrpc/src/*"],
+      "@defichain/jellyfish-block/*": ["jellyfish-block/src/*"],
```
Comment (canonbrother) [inline tsconfig.json:28]: FYI, the `/*` will be needed to be removed
https://github.com/DeFiCh/jellyfish/pull/658

File: packages/jellyfish-block/__tests__/rpc/blockHeader.test.ts (hunk 9 above)
Comment (Jouzo) [reply packages/jellyfish-block/__tests__/rpc/blockHeader.test.ts:24]: https://github.com/DeFiCh/ain/issues/737

**Review by canonbrother — APPROVED** [review]

**Review by surangap — APPROVED** [review]

**Review by jingyi2811 — APPROVED** [review]

**Review by fuxingloh — APPROVED** [review]

**Pull request merged**